### PR TITLE
Auct1037 rerequest email confirmation

### PR DESCRIPTION
**Pull request by erikdstock**
Jira Ticket [AUCT-1037](https://artsyproduct.atlassian.net/browse/AUCT-1037)
This PR duplicates #5667 by changing the static 'link expired' message to one which includes a mutation for re-requesting email confirmation. It is mostly copy-pasted over with a new mutation and analytics schema value.

Notably this banner, when triggered by `?flash_message=expired_token` will appear whether you _can_ request email confirmation (in a `me.canRequestEmailConfirmation` sense) or not.

Including the mobile view as an edge case:
image

TODO:

- [x] Take a stab at extracting the common email confirmation request component
- [x] Add type signature to the `selectContent` function

### CODE REVIEW

File: src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import { EmailConfirmationLinkExpiredMutationResponse } from "v2/__generated__/EmailConfirmationLinkExpiredMutation.graphql"
+import {
+  AnalyticsSchema as Schema,
+  useSystemContext,
+  useTracking,
+} from "v2/Artsy"
+
+import createLogger from "v2/Utils/logger"
+
+const logger = createLogger("EmailConfirmationLinkExpired")
```
Comment (damassi) [inline src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx:13]: Looking around at other use cases for `createLogger` we generally include the path

File: src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+              }
+            }
+          `,
+        })
+      }
+    )
+  }
+
+  const handleSubmit = () => {
+    logger.warn("click")
```
Comment (damassi) [inline src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx:55]: stray

File: src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    trackEvent({
+      action_type: Schema.ActionType.Click,
+      subject: Schema.Subject.EmailConfirmationLinkExpired,
+    })
+    logger.warn("track")
```
Comment (damassi) [inline src/v2/Components/FlashBanner/EmailConfirmationLinkExpired.tsx:60]: stray

File: src/v2/Components/FlashBanner/index.tsx [hunk 4]
```diff
@@ -49,10 +49,15 @@ export const FlashBanner: React.FC<FlashBannerProps> = props => {
...
-    content = contentMap[contentCode]
+  switch (contentCode) {
+    case "email_confirmation_cta":
+      content = <EmailConfirmationCTA />
+      break
+    case "expired_token":
+      content = <EmailConfirmationLinkExpired />
+      break
+    default:
+      content = contentMap[contentCode]
```
Comment (damassi) [inline src/v2/Components/FlashBanner/index.tsx:60]: Nice 👍 

This is a pattern that I often use to make make this a little less verbose:

```jsx
const getContent = () => {
  switch (contentCode) {
    case "email_confirmation_cta":
      return  <EmailConfirmationCTA />	
    case "expired_token":    
      return <EmailConfirmationLinkExpired />
    default:
      return contentMap[contentCode]
}

const content = getContent() 
```

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 5]
```diff
@@ -115,7 +115,93 @@ describe("FlashBanner", () => {
...
+        data: {
+          me: { canRequestEmailConfirmation: true },
+        },
+        queryString: "?flash_message=expired_token",
+      })
+
+      expect(wrapper.text()).toContain("Link expired.")
+    })
+
+    it.only("user seeing banner can click to re-trigger email confirmation message", async () => {
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:135]: stray `only`

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 6]
```diff
@@ -115,7 +115,93 @@ describe("FlashBanner", () => {
...
+  describe("user can request email confirmation (me?.canRequestEmailConfirmation=true)", () => {
+    it("user is prompted to re-request email confirmation if they can", async () => {
+      const { wrapper } = await getRelayWrapper({
+        data: {
+          me: { canRequestEmailConfirmation: true },
+        },
+        queryString: "?flash_message=expired_token",
+      })
+
+      expect(wrapper.text()).toContain("Link expired.")
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:132]: Can we assert upon the larger text that's now displayed. namely "Resend verification email"

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 7]
```diff
@@ -115,7 +115,93 @@ describe("FlashBanner", () => {
...
+                message: "Something Bad",
+              },
+            },
+          },
+        },
+      })
+
+      expect(wrapper.text()).toContain("Link expired.")
+
+      wrapper.find("button").first().prop("onClick")({} as any)
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:199]: simulate("click")

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 8]
```diff
@@ -161,10 +247,7 @@ describe("Email Confirmation CTA", () => {
 
       expect(wrapper.text()).toContain("Please verify your email address")
 
-      wrapper
-        .find("button")
-        .first()
-        .prop("onClick")({} as any)
+      wrapper.find("button").first().prop("onClick")({} as any)
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:250]: simulate("click")

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 9]
```diff
@@ -215,10 +295,7 @@ describe("Email Confirmation CTA", () => {
...
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:298]: simulate("click")

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx [hunk 10]
```diff
@@ -115,7 +115,93 @@ describe("FlashBanner", () => {
...
+            confirmationOrError: {
+              unconfirmedEmail: "ceo@blackwater.biz",
+            },
+          },
+        },
+      })
+
+      expect(wrapper.text()).toContain("Link expired.")
+
+      wrapper.find("button").first().prop("onClick")({} as any)
```
Comment (dleve123) [inline src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx:152]: simulate("click")

File: src/v2/Components/FlashBanner/index.tsx (hunk 4 above)
Comment (damassi) [reply src/v2/Components/FlashBanner/index.tsx]: what if you needed, say, a value from a hook? Then one would have to add arguments and pass things around, etc. But that said, its no horible to moveout of function body, just not strictly necessary. Its also nice to read a component top down and see exactly whats going on, versus having to jump all around. This falls apart of course if the component is more complex / longer, but for something like this, keeping it nice and dense increases readability.

File: src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx (hunk 7 above)
Comment (dleve123) [reply src/v2/Components/FlashBanner/__tests__/FlashBanner.jest.tsx]: Oh interesting - I wasn't aware of this! Thank you for clarifying – I read the issue and looks like there's a fair bit of back and forth on the semantics of shallow rendering vs. full DOM rendering with respect to the simulate API.

Regardless, I think, on principle, being consistent is more important than preparing for this API to be deprecated. A find+replace across the codebase when the API does indeed become deprecated should be pretty straightfoward.

**Pull request merged**